### ISSUE TITLE
Improvements and fixes in the standard documentation

### ISSUE THREAD

**Issue by winitzki** [dhall-lang/dhall-lang#1363]
I've been going through the Dhall standard and re-implementing everything from scratch ([in Scala](https://github.com/winitzki/scall)). In general I find the [standard documentation](https://github.com/dhall-lang/dhall-lang/blob/master/standard/) to be of excellent quality. I was able to implement the parser and the normalization reasonably easily. I just found some minor mismatches and omissions that sometimes got me confused or even made my tests fail.

Most of the mismatches happen because `beta-normalization.md` repeats the same algorithm first in a mathematical notation and then in example code in Haskell. In a few cases, the Haskell code does not fully repeat all the steps shown in the mathematical notation: either the final beta-normalization step is not shown, or some other steps are not shown. In one case it seems that the final beta-normalization step is completely missing (but required).

Here is a list of issues I have found so far (parser, alpha-normalization, beta-normalization).

1. The Syntax.hs data structures make a distinction between `Builtin` and `Constant`. However, the parser (`dhall.abnf`) does not make that distinction clearly. For example, the parser documentation says that a `nonreserved_label` cannot be a `Builtin`. It also cannot be a `Constant`, actually. The list of reserved identifiers is (correctly) a union of builtins and constants but the parser must parse `Natural` into a `Builtin` but `True` into a `Constant`. This was confusing to me.
2. One of the tests requires to parse an import of the form `./path0/path1/path2/file"`. At some point, my parser implementation had a bug where it disallowed any decimal digits in the paths other than `0` and `9`. (I forgot the dash in the character pattern `0-9`.) But all tests passed except for importing `./path0/path1/path2/file"`. I suggest adding more tests so that it is verified that the parser accepts (at least all ASCII) valid path characters and rejects invalid ones.
3. In substitution.md the code following `substitute (Lambda x _A₀ b₀) x' n e₀ | x == x' = Lambda x _A₁ b₁` is hard to follow because there are two cases with different variable names (`x`, `x'`, `y`). I would suggest rewriting like this:
```
substitute (Lambda y _A₀ b₀) x n e₀ | y == x = Lambda x _A₁ b₁
  where
    _A₁ = substitute _A₀ x n e₀

    e₁ = shift 1 y 0 e₀

    b₁ = substitute b₀ x(1 + n) e₁

substitute (Lambda y _A₀ b₀) x n e₀ = Lambda y _A₁ b₁
  where
    _A₁ = substitute _A₀ x n e₀

    e₁ = shift 1 y 0 e₀

    b₁ = substitute b₀ x n e₁
```
and similarly for `Forall` and `Let`. This also helps simplify the code because the only difference is `n + 1` vs. `n` in the last `substitute` function.

All the rest concerns `beta-normalization.md`.

4. The `Text/show` rules do not specify that characters like `\u0007` or `\u0010` must be escaped. But the test `TextShowAllEscapesA.dhall` fails unless those are escaped. Which characters really need to be escaped by `Text/show`?
5. The `Text/show` algorithm adds quotes around the string at the last step, but this is not shown in `beta-normalization.md`'s Haskell code (it is shown in the math notation).
6. Does `betaNormalize(toMap {=} T)` give `[] : U` where `U = betaNormalize T`? Or does it not need to normalize `T`? `beta-normalization.md` specifies `EmptyList _T₀`, which implies no normalization in that case, but I would expect it to be normalized.
7. When normalizing `RecordLiteral /\ RecordLiteral` with either `CombineRecordTerms` or `CombineRecordTypes`, the Haskell code does not include the final normalization step (but the math notation does). 
8. beta-normalization.md does not include the rule for `ls₀ === rs₀`  in Haskell code after `betaNormalize (Operator ls₀ Prefer rs₀)` .
9. When normalizing `Field(x // RecordLiteral, ...)` the final result may be sometimes a `Field` that again needs to be normalized. The Haskell code does not contain that step.
10. When normalizing `Field(x /\ RecordLiteral, ...)` or `Field(RecordLiteral /\ x)`, the final result must be again a `Field(...)` but the Haskell code says it's an `Operator CombineRecordTerms`. The math notation correctly applies a field selection.
11. When normalizing `ProjectByLabels(...)` the final result is `t1` and it still needs to be beta-normalized after `t1 = ProjectByLabels .(RecordLiteral rs) (filter predicate xs₀)` but the Haskell code does not do that.
12.   When normalizing `ProjectByType(RecordType ...)` the final result is a `ProjectByLabels` and still needs to be normalized.

I'll be happy to provide more details and point to specific places in the documentation if it's hard to find for whoever wants to look at it.

**Comment by Gabriella439** [comment]
Yeah, if you put up pull requests to amend the standard we'd be happy to merge them.  Even stuff that refactors the Haskell implementation to be more clear or to better match the standard is fair game.

**Comment by winitzki** [comment]
I will make a PR later after gathering some more nits and corrections to the standard. Right now I'm going through implementing type-checking and imports, and I could propose corrections there as well.

One question I have so far: Why is `Optional` a built-in type rather than just a type alias defined in the Prelude? I see that a builtin `List/head` returns an `Optional` but it could just as well return a literal union type `< None | Some A >`. I didn't find an explanation for this design choice in the documentation.

**Comment by MonoidMusician** [comment]
As I understand it, one of the reasons for `Optional` being a built-in is so that it is possible for output formats to handle it differently, such as going to `null` or `undefined` values (or missing keys in JSON) instead of having to be encoded like other sum types are. The reality is that most output formats have some kind of null value which is helpful to reference directly.

Also the type inference is much nicer for writing `Some` and `None`.
